### DISCUSSION TITLE
Able to run with  "nrsmpi turbpipe 1" but not  "nrsmpi turbpipe 2/4/8"

### DISCUSSION THREAD

**Discussion by tonyzahtila** [Nek5000/nekRS#250] [General]
Hi, 

  I've managed to successfully run the turbPipe case for 1 CPU/GPU. It looks pretty fast.

  However, I am now trying to run the case with 4 CPU/GPUs. I tend to execute the following commands in succession:
    
             nrspre turbPipe 4
             nrsmpi turbPipe 4

  The first thing that I will note is that whilst nrspre runs succesfully, it reports that it is only using 1 MPI task. Despite the fact that I have provided it with 4 MPI ranks and 4 GPUs. It does state it's jit-compiling for > 4 MPI tasks so maybe this is by design. I just found it confusing given that on the github front page it tells you to run for example, nrspre ethier 2.. (This isn't the case with nrsmpi)

                  ____   ___   / /__ / __ \/ ___/
                  / __ \ / _ \ / //_// /_/ /\__ \
                 / / / //  __// ,<  / _, _/___/ /
                /_/ /_/ \___//_/|_|/_/ |_|/____/  v20.1
                
                COPYRIGHT (c) 2019-2021 UCHICAGO ARGONNE, LLC
                    
                MPI tasks: 1
                
                using OCCA_DIR: /home/tzahtila/.local/nekrs/
                using OCCA_CACHE_DIR: /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/occa/
                
                Initializing device
                active occa mode: CUDA
                
                performing dry-run to jit-compile for >4 MPI tasks ...
                
                Building parallel face connectivity ... done (0.000209102s)
                Nq: 8 cubNq: 11

The issue I run into when running with **nrsmpi turbPipe 2** :

              calling nek_userchk ...
              loading mesh from nek ... NboundaryIDs: 3, NboundaryFaces: 1656 done (0.000588247s)
              Building parallel face connectivity ... done (0.00788698s)
              Nq: 8 cubNq: 11
              J in range [5.23098e-05,0.00066896]
                 setvert3d:   8          864037         2263717          864037          864037
              min NinternalElements: 3132 (ratio: 0.97)
              loading udf kernels ... done
              loading gs kernels ... done (0.948538s)
              timing oogs modes: 0.000545539s 0.000544805s 0.000548324s 0.000547695s used mode: 1.1
              High pass filter relaxation: chi = 8.0000 using 2 mode(s)
              loading ns kernels ... done (0.196599s)
              copying solution from nek
              calling udf_setup ... [spartan-gpgpu006:25914:0:25914] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x2b06d3eee010)
              ==== backtrace (tid:  25914) ====
               0 0x0000000000050a75 ucs_debug_print_backtrace()  ???:0
               1 0x0000000000264fc4 make_topology_unique()  /data/cephfs/punim0524/nek/nekRS/build/_deps/gs_content-src/src/gs.c:290
               2 0x0000000000268545 gs_unique()  /data/cephfs/punim0524/nek/nekRS/build/_deps/gs_content-src/src/gs.c:1461
               3 0x0000000000268545 gs_unique()  /data/cephfs/punim0524/nek/nekRS/build/_deps/gs_content-src/src/gs.c:1462
               4 0x000000000017a575 ogsGsUnique()  /data/cephfs/punim0524/nek/nekRS/3rd_party/gslib/ogs/src/ogsHostSetup.c:97
               5 0x0000000000184275 ogsSetup()  /data/cephfs/punim0524/nek/nekRS/3rd_party/gslib/ogs/src/ogsSetup.cpp:123
               6 0x00000000000aa649 velRecycling::setup()  /data/cephfs/punim0524/nek/nekRS/src/plugins/velRecycling.cpp:135
               7 0x000000000000128e UDF_Setup()  /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/udf/udf.cpp:29
               8 0x0000000000073499 nrsSetup()  /data/cephfs/punim0524/nek/nekRS/src/core/setup.cpp:546
               9 0x0000000000057614 nekrs::setup()  /data/cephfs/punim0524/nek/nekRS/src/lib/nekrs.cpp:101
              10 0x0000000000409aab main()  /data/cephfs/punim0524/nek/nekRS/src/main.cpp:130
              11 0x0000000000022555 __libc_start_main()  ???:0
              12 0x0000000000409fd7 _start()  ???:0
              =================================
              [spartan-gpgpu006:25914] *** Process received signal ***
              [spartan-gpgpu006:25914] Signal: Segmentation fault (11)
              [spartan-gpgpu006:25914] Signal code:  (-6)
              [spartan-gpgpu006:25914] Failing at address: 0x2c9f0000653a
              [spartan-gpgpu006:25914] [ 0] /lib64/libpthread.so.0(+0xf630)[0x2b02c64bb630]
              [spartan-gpgpu006:25914] [ 1] /home/tzahtila/.local/nekrs/lib/libnekrs.so(+0x264fc4)[0x2b02c58dbfc4]
              [spartan-gpgpu006:25914] [ 2] /home/tzahtila/.local/nekrs/lib/libnekrs.so(gs_unique+0x75)[0x2b02c58df545]
              [spartan-gpgpu006:25914] [ 3] /home/tzahtila/.local/nekrs/lib/libnekrs.so(ogsGsUnique+0x65)[0x2b02c57f1575]
              [spartan-gpgpu006:25914] [ 4] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_Z8ogsSetupiPxRP19ompi_communicator_tiN4occa6deviceE+0x295)[0x2b02c57fb275]
              [spartan-gpgpu006:25914] [ 5] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_ZN12velRecycling5setupEP5nrs_tN4occa6memoryExid+0x179)[0x2b02c5721649]
              [spartan-gpgpu006:25914] [ 6] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/udf/libUDF.so(UDF_Setup+0xae)[0x2b02ef97628e]
              [spartan-gpgpu006:25914] [ 7] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_Z8nrsSetupP19ompi_communicator_tN4occa6deviceER9setupAideP5nrs_t+0xa919)[0x2b02c56ea499]
              [spartan-gpgpu006:25914] [ 8] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_ZN5nekrs5setupEP19ompi_communicator_tiiiNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES7_S7_S7_+0x10e4)[0x2b02c56ce614]
              [spartan-gpgpu006:25914] [ 9] /home/tzahtila/.local/nekrs/bin/nekrs[0x409aab]
              [spartan-gpgpu006:25914] [10] /lib64/libc.so.6(__libc_start_main+0xf5)[0x2b02c66ea555]
              [spartan-gpgpu006:25914] [11] /home/tzahtila/.local/nekrs/bin/nekrs[0x409fd7]
              [spartan-gpgpu006:25914] *** End of error message ***
              -------------------------------------------------------
              Primary job  terminated normally, but 1 process returned
              a non-zero exit code. Per user-direction, the job has been aborted.

The issue I run into when running with **nrsmpi turbPipe 4:**

            loading ns kernels ... done (0.208494s)
            copying solution from nek
            calling udf_setup ... done
            ================ ELLIPTIC SETUP VELOCITY ================
            bID 1 -> bcType fixedValue
            bID 2 -> bcType zeroGradient
            bID 3 -> bcType zeroValue
            allNeumann = 0
            [spartan-gpgpu006:30457:0:30457] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x13469000)
            [spartan-gpgpu006:30458:0:30458] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x1391d008)
            [spartan-gpgpu006:30460:0:30460] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x134ef000)
            -------------------------------------------------------
            Primary job  terminated normally, but 1 process returned
            a non-zero exit code. Per user-direction, the job has been aborted.
            -------------------------------------------------------
            --------------------------------------------------------------------------
            mpirun noticed that process rank 3 with PID 30460 on node spartan-gpgpu006 exited on signal 11 (Segmentation fault).
            --------------------------------------------------------------------------
            

The issue I run into when running with **nrsmpi turbPipe 8:**
                       
                     done :: partioning   0.18088     sec
                    
                     reading mesh
                     reading curved sides
                     reading bc for ifld           1
                    [spartan-gpgpu006:12340:0:12340] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xffffffffffffff90)
                    [spartan-gpgpu008:22579:0:22579] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xffffffffffffff90)
              ==== backtrace (tid:  12340) ====
               0 0x0000000000050a75 ucs_debug_print_backtrace()  ???:0
               1 0x000000000002695e ucp_request_free()  ???:0
               2 0x0000000000007955 mca_pml_ucx_precv_completion()  ???:0
               3 0x0000000000041dd7 ucp_eager_only_handler()  ???:0
               4 0x0000000000014eb5 uct_mm_iface_release_desc()  ???:0
               5 0x000000000002cd6a ucp_worker_progress()  ???:0
               6 0x0000000000004344 mca_pml_ucx_progress()  ???:0
               7 0x0000000000032334 opal_progress()  ???:0
               8 0x000000000001aa12 ompi_osc_pt2pt_unlock()  ???:0
               9 0x000000000009d1bd PMPI_Win_unlock()  ???:0
              10 0x000000000005687b pmpi_win_unlock()  ???:0
              11 0x0000000000189498 dprocmapget_()  /home/tzahtila/.local/nekrs/nek5000/core/dprocmap.f:108
              12 0x0000000000189319 gllnid_()  /home/tzahtila/.local/nekrs/nek5000/core/dprocmap.f:176
              13 0x00000000001714dc readp_re2_bc_()  /home/tzahtila/.local/nekrs/nek5000/core/reader_re2.f:330
              14 0x000000000017103a read_re2_data_()  /home/tzahtila/.local/nekrs/nek5000/core/reader_re2.f:47
              15 0x00000000000329a5 nekf_setup_()  /home/tzahtila/.local/nekrs/nekInterface/nekInterface.f:200
              16 0x00000000000b0678 nek_setup()  /data/cephfs/punim0524/nek/nekRS/src/nekInterface/nekInterfaceAdapter.cpp:592
              17 0x000000000007be73 nrsSetup()  /data/cephfs/punim0524/nek/nekRS/src/core/setup.cpp:63
              18 0x0000000000057614 nekrs::setup()  /data/cephfs/punim0524/nek/nekRS/src/lib/nekrs.cpp:101
              19 0x0000000000409aab main()  /data/cephfs/punim0524/nek/nekRS/src/main.cpp:130
              20 0x0000000000022555 __libc_start_main()  ???:0
              21 0x0000000000409fd7 _start()  ???:0
              =================================
              [spartan-gpgpu006:12340] *** Process received signal ***
              [spartan-gpgpu006:12340] Signal: Segmentation fault (11)
              

Any help really appreciated,

Tony

**Comment by yslan** [comment]
Hi Tony,

nrspre is a script calling nekrs in the build-only mode running on single processor. 
It will compile the OCCA and Nek5000 in advance for the target number of GPU.
For example, `nrspre turbPipe 4` means your target number of GPU is 4 or larger.
 "`            performing dry-run to jit-compile for >4 MPI tasks ...`"

Therefore, at runtime, it will only need to load the kernels instead of compiling a lots of stuffs.
This will avoid the conflict that many processors compile the same things at the same time.

I suggest you first test the code using single node in the interactive mode.
In that case, you should be able to use nrspre and nrsmpi directly.

For many node, you have to modify the the submit script.
Calling nrsmpi might not be suitable. Please refer to `nrsqsub_summit`. 
You might also need system admin's help to bind GPU to CPU and optimize the UCX for communication on your network.

@aprilnovak does have a working in progress documentation for NekRS. I'm not sure if it's ready for public. 

Hope this helps,
Yu-Hsiang

**Reply by tonyzahtila** [reply]
Hey Yu-Hsiang,

Thank you for taking a look at my problem. You have cleared up proper running of nrspre for me.

However, I am not attempting to run on multiple nodes at this stage, only on one node, but would like to call 2/4 GPUs on that node. So I am at a loss why it won't run.

Tony

**Reply by yslan** [reply]
Hi Tony,

Do you have GPUs on the login node? or can you use interactive mode?

It'd be easier to figure out errors in a bash environment. 
Otherwise, you might mix another dozens of potential issues related to the submit commands/scripts.

Also, for debugging, I recommand setting verbose = true in the `.par` file like this
```
[GENERAL]
verbose = true
```
This should tell more details and help locates the error.

My guess is NekRS still trying to compile at the runtime.

Make sure you did `nrspre turbPipe 1` first.
Also, make sure the `OCCA_CACHE_DIR` printed from the nrspre matches the one printed at the runtime.

Yu-Hsiang

**Reply by tonyzahtila** [reply]
I am able to request an interactive node,

`sinteractive -p gpgpu -A punim0524 --gres=gpu:2 --ntasks=2`

I have tried this setting verbose = true in my .par file and  OCCA_CACHE_DIR from nrspre indeed matches that printed at runtime.

This is the complete error message I get, 

```
COPYRIGHT (c) 2019-2021 UCHICAGO ARGONNE, LLC

MPI tasks: 2

using OCCA_DIR: /home/tzahtila/.local/nekrs/
using OCCA_CACHE_DIR: /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/occa/

Initializing device
active occa mode: CUDA

loading nek ...
 Reading /data/gpfs/projects/punim0524/big_pipe/turbPipe/turbPipe.re2                                                                        
 WARNING: lgmres might be too low!
 nelgt/nelgv/lelt:       45951       45951       45951
 lx1/lx2/lx3/lxd:            8           8           8           1
 
 partioning elements to MPI ranks
 reading mesh 
 reading bc for ifld           1
[spartan-gpgpu052:27489:0:27489] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xffffffffffffff90)
==== backtrace (tid:  27489) ====
 0 0x0000000000050a75 ucs_debug_print_backtrace()  ???:0
 1 0x000000000002695e ucp_request_free()  ???:0
 2 0x0000000000007444 mca_pml_ucx_precv_completion()  ???:0
 3 0x0000000000041dd7 ucp_eager_only_handler()  ???:0
 4 0x0000000000014eb5 uct_mm_iface_release_desc()  ???:0
 5 0x000000000002cd6a ucp_worker_progress()  ???:0
 6 0x00000000000036a7 mca_pml_ucx_progress()  ???:0
 7 0x000000000002e44c opal_progress()  ???:0
 8 0x000000000001676d ompi_osc_pt2pt_unlock_internal()  osc_pt2pt_passive_target.c:0
 9 0x00000000000904cf PMPI_Win_unlock()  ???:0
10 0x000000000005565e ompi_win_unlock_f()  ???:0
11 0x000000000018af38 dprocmapget_()  /home/tzahtila/.local/nekrs/nek5000/core/dprocmap.f:108
12 0x000000000018adb9 gllnid_()  /home/tzahtila/.local/nekrs/nek5000/core/dprocmap.f:176
13 0x0000000000172f4c readp_re2_bc_()  /home/tzahtila/.local/nekrs/nek5000/core/reader_re2.f:330
14 0x0000000000172aaa read_re2_data_()  /home/tzahtila/.local/nekrs/nek5000/core/reader_re2.f:47
15 0x000000000009ee0f set_proc_map_()  /home/tzahtila/.local/nekrs/nek5000/core/map2.f:588
16 0x000000000009e4b3 mapelpr_()  /home/tzahtila/.local/nekrs/nek5000/core/map2.f:42
17 0x00000000000329ad nekf_setup_()  /home/tzahtila/.local/nekrs/nekInterface/nekInterface.f:199
18 0x00000000000b0678 nek_setup()  /data/cephfs/punim0524/nek/nekRS/src/nekInterface/nekInterfaceAdapter.cpp:592
19 0x000000000007be73 nrsSetup()  /data/cephfs/punim0524/nek/nekRS/src/core/setup.cpp:63
20 0x0000000000057614 nekrs::setup()  /data/cephfs/punim0524/nek/nekRS/src/lib/nekrs.cpp:101
21 0x0000000000409aab main()  /data/cephfs/punim0524/nek/nekRS/src/main.cpp:130
22 0x0000000000022555 __libc_start_main()  ???:0
23 0x0000000000409fd7 _start()  ???:0
=================================
[spartan-gpgpu052:27489] *** Process received signal ***
[spartan-gpgpu052:27489] Signal: Segmentation fault (11)
[spartan-gpgpu052:27489] Signal code:  (-6)
[spartan-gpgpu052:27489] Failing at address: 0x2c9f00006b61
[spartan-gpgpu052:27489] [ 0] /lib64/libpthread.so.0(+0xf630)[0x2ac56f64f630]
[spartan-gpgpu052:27489] [ 1] /lib64/libucp.so.0(ucp_request_free+0xe)[0x2ac573ce695e]
[spartan-gpgpu052:27489] [ 2] /usr/local/easybuild/software/OpenMPI/3.1.0-GCC-6.2.0-cuda9.2/lib/openmpi/mca_pml_ucx.so(mca_pml_ucx_precv_completion+0xd4)[0x2ac573ca6444]
[spartan-gpgpu052:27489] [ 3] /lib64/libucp.so.0(ucp_eager_only_handler+0x2a7)[0x2ac573d01dd7]
[spartan-gpgpu052:27489] [ 4] /usr/lib64/libuct.so.0(+0x14eb5)[0x2ac584015eb5]
[spartan-gpgpu052:27489] [ 5] /lib64/libucp.so.0(ucp_worker_progress+0x5a)[0x2ac573cecd6a]
[spartan-gpgpu052:27489] [ 6] /usr/local/easybuild/software/OpenMPI/3.1.0-GCC-6.2.0-cuda9.2/lib/openmpi/mca_pml_ucx.so(mca_pml_ucx_progress+0x17)[0x2ac573ca26a7]
[spartan-gpgpu052:27489] [ 7] /usr/local/easybuild/software/OpenMPI/3.1.0-GCC-6.2.0-cuda9.2/lib/libopen-pal.so.40(opal_progress+0x2c)[0x2ac57199744c]
[spartan-gpgpu052:27489] [ 8] /usr/local/easybuild/software/OpenMPI/3.1.0-GCC-6.2.0-cuda9.2/lib/openmpi/mca_osc_pt2pt.so(+0x1676d)[0x2ac573fee76d]
[spartan-gpgpu052:27489] [ 9] /usr/local/easybuild/software/OpenMPI/3.1.0-GCC-6.2.0-cuda9.2/lib/libmpi.so.40(PMPI_Win_unlock+0xbf)[0x2ac56ea5e4cf]
[spartan-gpgpu052:27489] [10] /usr/local/easybuild/software/OpenMPI/3.1.0-GCC-6.2.0-cuda9.2/lib/libmpi_mpifh.so.40(ompi_win_unlock_f+0x1e)[0x2ac56fc7f65e]
[spartan-gpgpu052:27489] [11] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(dprocmapget_+0x158)[0x2ac5da5fef38]
[spartan-gpgpu052:27489] [12] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(gllnid_+0x89)[0x2ac5da5fedb9]
[spartan-gpgpu052:27489] [13] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(readp_re2_bc_+0x37c)[0x2ac5da5e6f4c]
[spartan-gpgpu052:27489] [14] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(read_re2_data_+0x20a)[0x2ac5da5e6aaa]
[spartan-gpgpu052:27489] [15] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(set_proc_map_+0x25f)[0x2ac5da512e0f]
[spartan-gpgpu052:27489] [16] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(mapelpr_+0x373)[0x2ac5da5124b3]
[spartan-gpgpu052:27489] [17] /data/gpfs/projects/punim0524/big_pipe/turbPipe/.cache/nek5000/libturbPipe.so(nekf_setup_+0x28d)[0x2ac5da4a69ad]
[spartan-gpgpu052:27489] [18] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_Z9nek_setupP19ompi_communicator_tR9setupAideP5nrs_t+0x5e8)[0x2ac56ec5d678]
[spartan-gpgpu052:27489] [19] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_Z8nrsSetupP19ompi_communicator_tN4occa6deviceER9setupAideP5nrs_t+0x132f3)[0x2ac56ec28e73]
[spartan-gpgpu052:27489] [20] /home/tzahtila/.local/nekrs/lib/libnekrs.so(_ZN5nekrs5setupEP19ompi_communicator_tiiiNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES7_S7_S7_+0x10e4)[0x2ac56ec04614]
[spartan-gpgpu052:27489] [21] /home/tzahtila/.local/nekrs/bin/nekrs[0x409aab]
[spartan-gpgpu052:27489] [22] /lib64/libc.so.6(__libc_start_main+0xf5)[0x2ac56f87e555]
[spartan-gpgpu052:27489] [23] /home/tzahtila/.local/nekrs/bin/nekrs[0x409fd7]
[spartan-gpgpu052:27489] *** End of error message ***
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 27489 on node spartan-gpgpu052 exited on signal 11 (Segmentation fault).



```

**Reply by yslan** [reply]
I'm not sure if this is related to UCX. 
Is it possible for you to try openmpi without UCX? 

Without UCX, it will be slightly slow, but at least I managed to run multiple GPU without it.
At this stage, I just want to rule out some possible errors.

p.s. The following combination works for me months before.
  openmpi 4.0.5
  ucx 1.9.0  (1.9.0-rc1 fails)

**Comment by stgeke** [comment]
I think this is related to a bug in UCX: https://github.com/Nek5000/nekRS/issues/201

**Reply by zongzilin** [reply]
Has this bug been fixed? I am having the similar problem on nekRS-21.1 as
`[hpc217:264304] Read -1, expected 43016, errno = 14
[hpc217:264304] *** Process received signal ***
[hpc217:264304] Signal: Segmentation fault (11)
[hpc217:264304] Signal code: Invalid permissions (2)
[hpc217:264304] Failing at address: 0x2b30f1dbd800
[hpc217:264303] Read -1, expected 43016, errno = 14
[hpc217:264303] *** Process received signal ***
[hpc217:264303] Signal: Segmentation fault (11)
[hpc217:264303] Signal code: Invalid permissions (2)
`

**Comment by tonyzahtila** [comment]
I believe it was resolved by certain UCX settings on my machine. Granted,
we are running a supercomputer in Australia, so it may not be helpful for
you, but we include such settings in the slurm script:

export UCX_MEM_MMAP_RELOC=n
export UCX_MEM_MALLOC_HOOKS=n
export UCX_MEM_MALLOC_RELOC=n
export UCX_MEM_EVENTS=n
export UCX_MEMTYPE_CACHE=n

backend=CUDA

env | grep 'UCX\|OMPI' | sort
